### DISCUSSION TITLE
Custom properties on Collections vanish between render initiation and Blade view

### DISCUSSION THREAD

**Discussion by kokoshneta** [livewire/livewire#5000] [1. Help]
I have a Livewire component for a search bar that searches across different model types. I store (or would like to store) found hits in a `$results` property in the component.

The object returned from the search function and containing the search results is a Collection object containing an array of Eloquent Collection objects (one per model type), each containing an array of Eloquent models (the individual hits).

Each of the Eloquent Collections has some additional (public) properties, apart from the `$items` array, and look something like this (copy-pasted from a `dd()` dump):

```
^ Illuminate\Support\Collection {#1574 ▼
  #items: array:2 [▼
    0 => Illuminate\Database\Eloquent\Collection {#1703 ▼
      #items: array:5 [▶]
      #escapeWhenCastingToString: false
      +"route": "titles"
      +"line": 3
      +"type": 12
    }
    1 => Illuminate\Database\Eloquent\Collection {#1693 ▶}
  ]
  #escapeWhenCastingToString: false
}
```

When I do a search and store this object in `$this->results` in the component, it is correctly stored. If I `dd($this->results)` in the `render()` function immediately before returning the view, the output is that seen above, exactly as expected.

But if I `dd()` the object on the very first line in the Blade view, the object has changed and now looks like this:

```
^ Illuminate\Support\Collection {#1574 ▼
  #items: array:2 [▼
    0 => Illuminate\Database\Eloquent\Collection {#1703 ▼
      #items: array:5 [▶]
      #escapeWhenCastingToString: false
    }
    1 => Illuminate\Database\Eloquent\Collection {#1693 ▶}
  ]
  #escapeWhenCastingToString: false
}
```

The items remain the same, as does the protected `$escapeWhenCastingToString` property, but my public properties are nowhere to be seen and cannot be accessed. It’s as though only protected properties are brought through to the view.

The same issue was reported [on StackOverflow almost a year ago](https://stackoverflow.com/questions/68154792/how-can-i-send-property-which-added-to-collection-to-blade-in-livewire-component), but there are no answers on that question. I can’t find any other mention of the issue anywhere.

Is there some way to get the Collection in its entirety along to the view where I need it?

**Comment by kokoshneta** [comment]
I have since found a solution that works in my case, though it wouldn’t necessarily work for everyone. My `render()` method looked like this:

```
public function render() {
    if ($this->query) {
        $this->results = Searcher::search($this->query);
    }

    return view('livewire.searchbox');
}
```

If I change this around to pass the results collection returned from the search directly, without saving it in the component, everything works and the object is passed directly to the view with no change. Since the search results are retrieved anew for every render anyway, they don’t need to persist and it’s not really necessary to save them in the component, so that works out fine in this particular case:

```
public function render() {
    $results = $this->query ? Searcher::search($this->query) : NULL;
    return view('livewire.searchbox', ['results' => $results);
}
```

I still don’t have a clue why saving it in the component is not possible, though.

The [documentation](https://laravel-livewire.com/docs/2.x/properties#important-notes) notes that only a small subset of PHP object types will persist, including Laravel Collections, but it seems Collections _do not_ actually persist, at least not without being altered first (presumably through some sort of class reflection?).

**Comment by joshhanley** [comment]
@kokoshneta Livewire uses Laravel's queue serialiser under the hood, and it can serialise Eloquent Collections (but not grouped or nested) or standard collections (even with collections inside of it), but it doesn't know how to handle Models or Eloquent Collections inside of a standard collection, so instead they get converted to arrays. That is also why your custom properties are missing from your collections.

You can either use computed properties to generate your collection (rather than saving to a public property) or just put it in the render method like you have.

Hope this helps!
